src/scheduler/pbs_sched.c

```diff
@@ -792,45 +506,424 @@ are_we_primary()
+	while (servers[i] != NULL) i++;
+	servers[i] = svr;
+	log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO, msg_daemonname, "Connected to server %s", svrhost);
+}
+
+/**
+ * @brief Connect to all configured servers and populate servers list with
+ *        connections to servers
+ *
+ * @return void
+ */
+static void
+connect_servers(void)
+{
+	int max_connection = 1;
+
+	if (poll_context == NULL) {
+		poll_context = tpp_em_init(max_connection);
+		if (poll_context == NULL) {
+			log_errf(-1, __func__, "Failed to init cmd connections context");
+			die(-1);
+		}
+
+		servers = (sched_svrconn **) calloc(1, sizeof(sched_svrconn *) * max_connection + 1);
+		if (servers == NULL) {
+			log_errf(errno, __func__, MEM_ERR_MSG);
+			die(-1);
+		}
+
+		sched_cmds = new_ds_queue();
+		if (sched_cmds == NULL) {
+			log_errf(errno, __func__, MEM_ERR_MSG);
+			die(-1);
+		}
+	}
+
+	connect_server(pbs_default());
+}
+
+/**
+ * @brief reconnect to given server
+ *
+ * @param[in] sconn - pointer server struct to reconnect
+ *
+ * @return void
+ */
+static void
+reconnect_server(sched_svrconn *sconn)
+{
+	close_server(sconn);
+	connect_server(sconn->svrhost);
+	free(sconn->svrhost);
+	free(sconn);
+}
+
+/**
+ * @brief find server structure based on given secondary connection
+ *
+ * @param[in] sock - secondary connection to server
+ *
+ * @return sched_svrconn *
+ * @retval NULL  - failure
+ * @retval !NULL - success
+ */
+static sched_svrconn *
+find_server(int sock)
+{
+	int i = 0;
+	for (; servers[i] != NULL && servers[i]->secondary_sock == sock; i++)
+		return servers[i];
+	return NULL;
+}
+
+/**
+ * @brief read incoming command from given secondary connection
+ *        and add it into sched_cmds array
+ *
+ * @param[in]  sock   - secondary connection to server
+ *
+ * @return int
+ * @retval -2 - failure due to memory optation failed
+ * @retval -1 - failure while reading command
+ * @return  0 - no cmd, server might have closed connection
+ * @return  1 - success, read atleast one command
+ */
+static int
+read_sched_cmd(int sock)
+{
+	int rc = -1;
+	sched_cmd *cmd;
+
+	cmd = malloc(sizeof(sched_cmd));
+	if (cmd == NULL)
+		return -2;
+
+	rc = get_sched_cmd(sock, cmd);
+	if (rc != 1) {
+		free(cmd);
+		return rc;
+	} else {
+		sched_cmd *cmd_prio;
+
+		if (!ds_enqueue(sched_cmds, cmd)) {
+			if (cmd->jid)
+				free(cmd->jid);
+			free(cmd);
+			/* return error if we fail to enqueue first command */
+			return -2;
+		}
+
+		/*
+		 * There is possibility that server has sent
+		 * priority command after first non-priority command,
+		 * while we were in schedule()
+		 *
+		 * so try read it in non-blocking mode, but don't
+		 * return any failure if fails to read, as we have
+		 * successfully enqueued first command
+		 */
+		cmd_prio = malloc(sizeof(sched_cmd));
+		if (cmd_prio == NULL)
+			return 1; /* return 1 as we have enqueued first command */
+		cmd_prio->cmd = -1;
+		cmd_prio->jid = NULL;
+
+		get_sched_cmd_noblk(sock, cmd);
```

Comment:
I think it is better to test the return value of get_sched_cmd_noblk. It is because it might return 0 which indicates EOF i.e. server is closed.

Reply:
good point, missed that. will change

Reply:
@suresh-thelkar I tried adding check for 0 here and it always return after this call as for cmd_noblk, 0 means no cmd to read also not just EOF (for blocking version 0 means EOF). So I think we can't add check for 0 here.
And also I think we don't need any check for this call, as we are not returning any error once we reach this point due we already enqueued one command in cmds queue and from this point we are just checking whether prio command is there or not before we call to schedule()

Reply:
> @suresh-thelkar I tried adding check for 0 here and it always return after this call as for cmd_noblk, 0 means no cmd to read also not just EOF (for blocking version 0 means EOF). So I think we can't add check for 0 here.

The problem is zero being used for both 1) No command  and 2)EOF case. We can actually modify this function to return a separate return code for both these cases. Don't know whether we need to go to that extent.
I was thinking if we catch server going down here it is an early catch. Anyways because of the above reasons it is fine not do this for the time being I think.

Seems the following line of function header is wrong to me. I prefer to change it to as follows so that it is more clear. It is up to you.

* @retval	0	for EOF
We can change this to 
* @retval	0	if there is no super high priority command or EOF in case of server going down

Reply:
sure I will change func header to mention it.